### DISCUSSION TITLE
Need some help debugging PIU

### DISCUSSION THREAD

**Discussion by linfan68** [Moddable-OpenSource/moddable#1152] [Q&A]
I'm trying to add this OLED display driver: it's a 256x64, 16 level color OLED SPI display, very similar to ssd1306, just larger and supports gray level.

I write the test code like this(copied from the countdown demo), it runs well on ESP32-S3 + another ili9341 RGB SPI display.
```ts
import { Port } from "piu/MC"
import {} from "piu/MC"

declare global { function hsl(h: number, s: number, l: number): string }

const itemStyle = new Style({ font:"20px Open Sans", color:"white", horizontal:"left" });

class PortBehavior extends Behavior {
  onCreate(port: Port, delta) {
    trace(`onCreate\n`)
  }
  onDisplaying(port: Port) {
    trace(`onDisplaying\n`)
    port.interval = 2000
    port.start()
  }
  onDraw(port: Port, x: number, y: number, w: number, h: number) {
    trace(`onDraw\n`)
    port.fillColor('#202020', 0, 0, 256, 64)
    port.style = itemStyle;
    port.drawLabel('Hello World!!', 0, 0, 200, 20)

  }
  onTimeChanged(port) {
    trace(`onTimeChanged\n`)
    port.invalidate()
  }
}

let CountdownApplication = Application.template($ => ({
  skin: new Skin({ fill:"black" }),
  contents: [
    Port($, { width:256, height:64, Behavior:PortBehavior }),
  ]
}))

export default new CountdownApplication(null, { commandListLength:4096, displayListLength:4096, touchCount:0 })
```

to bring up sh1122, I added a driver (similar to ssd1306) with .js/.c/manifest.json files
the JS file looks like this:
```
import Bitmap from "commodetto/Bitmap";

export default class SH1122 @ "xs_SH1122_destructor" {
	constructor(dictionary) @ "xs_SH1122";

	begin(x, y, width, height) @ "xs_SH1122_begin";
	send(pixels, offset, count) @  "xs_SH1122_send";
	end() @  "xs_SH1122_end";

	continue() {
		// empty implementation overrides PixelOut.continue which throws		//@@ check this
	}
	pixelsToBytes(count) {
		return count;
	}

	get pixelFormat() {return Bitmap.Gray256;}
	get width() @ "xs_SH1122_width";
	get height() @ "xs_SH1122_height";

	get c_dispatch() @ "xs_SH1122_get_c_dispatch";

	adaptInvalid(r) {
		trace($`SH1122.adaptInvalid ${this.width}, ${this.height}\n`);
		r.x = 0;
		r.y = 0;
		r.w = this.width;
		r.h = this.height;
	}
}
```

on the C side, I got function xs_SH1122 working and successfully initialized the display.
But the PIU engine does not sending drawing command to SH1122 driver (like begin(), send(), end(), continue() )
It does ask for width and height (and I returned the correct non-zero value, otherwise the onDraw() function does not get called.

I got deeply stuck here... It must be some configuration error here... I'm trying to trace back the calling tree to figure out with branch makes it not sending any data to the driver.... but I found it's really hard to navigate through the code base, mostly because frequent jumps between native and JS code.

So my questions are:
* Is there some good way to trace the callstack across JS and C? let's say I switch to the ili9341 display and print the stack trace from C code, all I got was an interop call from xsHost. (I know it's hard, like tracing async callstack is hard...) 

* Any function or file I can put a bunch of printf or trace to debug my issue?

**Comment by phoddie** [comment]
How did you build your driver? I'm curious how different it is from the ssd1306 driver.

When I bring up a new display driver, I always start with Poco rather than Piu. It is a much simpler runtime to debug. Once that works, that I test with Piu. $MODDABLE/examples/commodetto/origin is a trivial place to start.

You can print a JavaScript stack from C using XS in C. As you know, the XS implements the  implements the proposed `stack` property of the `Error` class. So, create an instance of the JavaScript `Error` object in C using `xsNew`, get the `stack` property, get the value of the property as a string, and call `printf` or `xsTrace`. I don't have time to write that code out fully this morning, but if you need help please let me know. And if you get it working, please post the snippet so others can use it too.

**Reply by linfan68** [reply]
Thanks for your help! I’ll give it a try tonight. Sh1122 should be very similar to ssd1306. Some command addresses needs to be changed. Also there are other subtle differences, like ssd1306 uses uint8 to store width and height, but sh1122 has 256 px width. So the original code wraps back to zero width and does not render.

**Reply by linfan68** [reply]
@phoddie I did some searching in the code base but cannot find a good example to create xs Error object and print it... Also I found the xsMachine has a "stack" property, may be what I need is just a code snippet that prints it out? Currently I'm not comfortable pulling data from XS in C code...

BTW I figured out my problem and unblocked myself (with LOTS of printfs)
Basically, I didn't realize there is a "short-cut" "PixelsOutDispatchRecord" that allows c code directly call into display drivers' methods... so I keep looking at the path that `void xs_SH1122_begin(xsMachine *the)` didn't get called. But the underlying `SH1122Begin(ssd, xMin, yMin, width, height)` code is getting called correctly from the very beginning....

**Reply by phoddie** [reply]
> I did some searching in the code base but cannot find a good example to create xs Error object and print it... Also I found the xsMachine has a "stack" property, may be what I need is just a code snippet that prints it out? Currently I'm not comfortable pulling data from XS in C code...

XS in C is the way to go here. If you think about it as an API for implementing JavaScript operations from C, it becomes very straightforward. If you haven't read Chapter 11 of [our book](https://www.moddable.com/book), it is a step-by-step starting point for learning the basics of XS in C.

In this case, the JavaScript you would write is this:

```js
let e = new Error;
let s = e.stack;
trace(`${e}\n`);
```

Using XS in C, it is also just three lines, each equivalent to a line of JavaScript:

```c
#include "xsmc.h"
#include "mc.xs.h"			// for xsID_ values

void xs_traceJavaScriptStack(xsMachine * the)
{
	xsmcNew(xsResult, xsGlobal, xsID_Error, NULL);
	xsmcGet(xsResult, xsResult, xsID_stack);
	xsLog("%s\n", xsmcToString(xsResult));
}
```

Here's a small JavaScript test using the funciton:

```js
function traceJavaScriptStack() @ "xs_traceJavaScriptStack";

function a() {
	traceJavaScriptStack();
}

function b() {
	a();
}

b();
```

And the output:

```
Error
 at a (/Users/hoddie/Projects/moddable/examples/helloworld/main.js:4)
 at b (/Users/hoddie/Projects/moddable/examples/helloworld/main.js:8)
 at main (/Users/hoddie/Projects/moddable/examples/helloworld/main.js:11)
```

> BTW I figured out my problem and unblocked myself (with LOTS of printfs)
> Basically, I didn't realize there is a "short-cut" "PixelsOutDispatchRecord" that allows c code directly call into display drivers' methods... so I keep looking at the path that void xs_SH1122_begin(xsMachine *the) didn't get called. But the underlying SH1122Begin(ssd, xMin, yMin, width, height) code is getting called correctly from the very beginning....

Glad you were able to figure that out! The native dispatch is optional, but a really important performance optimization to eliminate the overhead of crossing the JavaScript/C language boundary many times while rendering a frame.

**Reply by phoddie** [reply]
@linfan68 – did this work out for you? Or maybe you didn't get a chance to try since you found the problem?
